flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NettyShuffleService.java

```diff
@@ -0,0 +1,124 @@
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * Netty based shuffle service implementation.
+ */
+public class NettyShuffleService implements ShuffleService<ResultPartition, SingleInputGate> {
+
+	private static final String METRIC_GROUP_NETWORK = "Network";
+	private static final String METRIC_TOTAL_MEMORY_SEGMENT = "TotalMemorySegments";
+	private static final String METRIC_AVAILABLE_MEMORY_SEGMENT = "AvailableMemorySegments";
+
+	@Override
+	public NettyShuffleEnvironment createShuffleEnvironment(ShuffleEnvironmentContext shuffleEnvironmentContext) {
+		NettyShuffleEnvironmentConfiguration networkConfig = NettyShuffleEnvironmentConfiguration.fromConfiguration(
+			shuffleEnvironmentContext.getConfiguration(),
+			shuffleEnvironmentContext.getMaxJvmHeapMemory(),
+			shuffleEnvironmentContext.isLocalCommunicationOnly(),
+			shuffleEnvironmentContext.getHostAddress());
+		return createNettyShuffleEnvironment(
+			networkConfig,
+			shuffleEnvironmentContext.getLocation(),
+			shuffleEnvironmentContext.getEventPublisher(),
+			shuffleEnvironmentContext.getParentMetricGroup(),
+			shuffleEnvironmentContext.getIOManager());
+	}
+
+	@VisibleForTesting
+	static NettyShuffleEnvironment createNettyShuffleEnvironment(
+			NettyShuffleEnvironmentConfiguration config,
+			ResourceID taskExecutorLocation,
+			TaskEventPublisher taskEventPublisher,
+			MetricGroup metricGroup,
+			IOManager ioManager) {
+		checkNotNull(ioManager);
+		checkNotNull(taskEventPublisher);
+		checkNotNull(config);
+
+		NettyConfig nettyConfig = config.nettyConfig();
+
+		ResultPartitionManager resultPartitionManager = new ResultPartitionManager();
+
+		ConnectionManager connectionManager = nettyConfig != null ?
+			new NettyConnectionManager(resultPartitionManager, taskEventPublisher, nettyConfig, config.isCreditBased()) :
+			new LocalConnectionManager();
+
+		NetworkBufferPool networkBufferPool = new NetworkBufferPool(
+			config.numNetworkBuffers(),
+			config.networkBufferSize(),
+			config.networkBuffersPerChannel());
+
+		registerNetworkMetrics(metricGroup, networkBufferPool);
+
+		ResultPartitionFactory resultPartitionFactory = new ResultPartitionFactory(
+			resultPartitionManager,
+			ioManager,
+			networkBufferPool,
+			config.networkBuffersPerChannel(),
+			config.floatingNetworkBuffersPerGate());
+
+		SingleInputGateFactory singleInputGateFactory = new SingleInputGateFactory(
+			taskExecutorLocation,
+			config,
+			connectionManager,
+			resultPartitionManager,
+			taskEventPublisher,
+			networkBufferPool);
+
+		return new NettyShuffleEnvironment(
+			taskExecutorLocation,
+			config,
+			networkBufferPool,
+			connectionManager,
+			resultPartitionManager,
+			resultPartitionFactory,
+			singleInputGateFactory);
+	}
+
+	private static void registerNetworkMetrics(MetricGroup metricGroup, NetworkBufferPool networkBufferPool) {
+		checkNotNull(metricGroup);
```

Comment:
might remove this if this check could be done before in `createNettyShuffleEnvironment`